### PR TITLE
Started printing out error messages when the observatory won't be reachable because of permissions problems.

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

iOS 14 has changed their policy for mdns permissions.  This provides some guidance if users run afoul of them.

## Related Issues

https://github.com/flutter/flutter/issues/60634
https://github.com/flutter/flutter/issues/63893

## Tests

I added the following tests:

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm [hunk 1]
```diff
@@ -113,6 +113,11 @@ - (void)publishServiceProtocolPort:(NSString*)uri {
 
   if (err != 0) {
     FML_LOG(ERROR) << "Failed to register observatory port with mDNS.";
+    if (@available(iOS 14.0, *)) {
+      FML_LOG(ERROR) << "Make sure the 'NSBonjourServices' key is set in your Info.plist for '"
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:117]: Can we point out it's only needed for Debug and Profile configuration `Info.plist`s?  At some point we should write up some info on the website, but in the meantime...

**Comment by jmagman** [comment]
`build_ios_debug_sim` didn't like `kDNSServiceErr_PolicyDenied`
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8870232824019334080/+/steps/build_ios_debug_sim/0/stdout
```
../../flutter/shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:135:27: error: use of undeclared identifier 'kDNSServiceErr_PolicyDenied'
  } else if (errorCode == kDNSServiceErr_PolicyDenied) {
```

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:117]: Sg, done.

**Comment by gaaclarke** [comment]
> `build_ios_debug_sim` didn't like `kDNSServiceErr_PolicyDenied`
> https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8870232824019334080/+/steps/build_ios_debug_sim/0/stdout
> 
> ```
> ../../flutter/shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:135:27: error: use of undeclared identifier 'kDNSServiceErr_PolicyDenied'
>   } else if (errorCode == kDNSServiceErr_PolicyDenied) {
> ```

I didn't see a guard in the header, but apparently it is iOS 14.0+.

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm (hunk 1 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:117]: Debug and Profile, not just Debug.

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm [hunk 2]
```diff
@@ -113,11 +113,25 @@ - (void)publishServiceProtocolPort:(NSString*)uri {
...
+                     << "(See also: https://developer.apple.com/news/?id=0oi77447)";
+    }
   } else {
     DNSServiceSetDispatchQueue(_dnsServiceRef, dispatch_get_main_queue());
   }
 }
 
+/// TODO(aaclarke): Remove this preprocessor macro once infra is moved to Xcode 12.
+static const DNSServiceErrorType kFlutter_DNSServiceErr_PolicyDenied =
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 140000
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:128]: Can you use a `@available` here?  I've needed to use this macro outside of the class (like for imports) but hopefully `@available` works inline?  Not a big deal though.

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm [hunk 3]
```diff
@@ -113,11 +113,25 @@ - (void)publishServiceProtocolPort:(NSString*)uri {
...
+    kDNSServiceErr_PolicyDenied;
+#else
+    // Found in usr/include/dns_sd.h.
+    -65570;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:132]: 👍

**Review by jmagman — APPROVED** [review]
LGTM with nit that the key should be added for both Debug and Profile.

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:128]: Na, that's why I messed it up the first time.  It's in an enum declaration that doesn't have annotations or guards of what version the operating system is.

**Pull request merged**